pkg/ddl/partition.go

```diff
@@ -3016,7 +3035,31 @@ func (w *worker) onReorganizePartition(d *ddlCtx, t *meta.Meta, job *model.Job)
 		// Should it actually be synchronous?
 		// Include the old table ID, if changed, which may contain global statistics,
 		// so it can be reused for the new (non)partitioned table.
-		asyncNotifyEvent(d, &util.Event{Tp: job.Type, TableInfo: tblInfo, PartInfo: statisticsPartInfo})
+
+		var event *statsutil.DDLEvent
+		switch job.Type {
+		case model.ActionReorganizePartition:
+			event = statsutil.NewReorganizePartitionEvent(
+				tblInfo,
+				statisticsPartInfo,
+				droppedPartInfo,
+			)
+		case model.ActionAlterTablePartitioning:
+			event = statsutil.NewAddPartitioningEvent(
+				tblInfo,
+				statisticsPartInfo,
+			)
+
+		case model.ActionRemovePartitioning:
+			event = statsutil.NewRemovePartitioningEvent(
+				tblInfo,
+				// TODO(hi-rustin): figure out why we need to pass statisticsPartInfo here.
+				statisticsPartInfo,
+			)
+		default:
+			return ver, errors.Errorf("unknown job type: %s", job.Type.String())
+		}
```

Comment:
Could we extract the above code as a function?

Reply:
Good suggestion! Updated.